Experimental MT

ysbaddaden opened the pull request:
Rework of #7214 with much better integration, and acceptable overall performance: -10% slower with `NPROCS=1` and `-Dmt` than with the current scheduler for a simple HTTP::Server stress test using `wrk`.

EDIT: despite achieving good single thread performance despite being thread safe, adding more threads won't improve performance. One of four threads will roughly be equal —allright, 4 threads is slightly faster than 1, but multiple processes are still much faster.

NOTE: I want clean commits for each feature and improvement. I'll force-push revised commits instead of pushing lots of fixups —I already squashed dozens of them.

ysbaddaden committed:
EventLoop: enable thread protections

- Enable support for Windows and POSIX threads in libevent2, so
  adding events to a single `Event::Base` from different threads
  is safe;

- Rename `Event::base.run` methods as `Event::Base#loop(flags)`
  method so callers can choose to block, run the loop once or to
  never block.

ysbaddaden committed:
Introduce Fiber::Context struct

The fiber context holds the current stack top pointer (for a saved
stack), as well as a resumable flag that tells whether the fiber can
be resumed (its context was fully saved) or is currently running, or
is dead (its proc returned).

The resumable flag is required to prevent a fiber to be resumed
while it's still running. This should usually not happen with the
monothreaded scheduler, but still useful to find a buggy
synchronisation primitive that would enqueue the same fiber multiple
times (e.g. the current `select` implementation). This will be very
frequent with a multithreaded scheduler where a thread B could try
to resume a fiber that has just been enqueued by thread A but didn't
fully store its context yet.

The fiber status, running or dead, is monitored using an `@alive`
boolean. We can't set `context.resumable` to another value, because
the scheduler eventually swaps the context to another fiber, which
would overwrite `context.resumable`.

The scheduler now aborts with a fatal error whenever it would resume
a running or dead fiber, which would otherwise lead to a segfault.
It doesn't just ignore the fiber since to enqueue a fiber twice or a
dead fiber is an error that must be fixed.

ysbaddaden committed:
Add Atomic::Flag

ysbaddaden committed:
MT: mutual exclusion lock (Mutex)

Replaces the Channel-based Mutex implementation with a SpinLock and
WaitQueue based one that is thread safe, fiber aware, and doesn't
allocate —unlike Channel and its internal reallocatable Deque.

Also implements the Crystal::SpinLock, Crystal::Mutex and
Crystal::ConditionVariable internal structs that will be necessary
to write thread-safe synchronisation primitives.

ysbaddaden committed:
MT: BDWGC safety

Uses a (multi)readers (single)writer lock to prevent BDWGC from
initiating a scan/collect while a scheduler thread is swapping
context, which could lead to report a mismatched stack pointer and
stack bottom for a thread, which would obviously segfault.

Implements a custom RWLock struct which is noticeably faster than
`pthread_rwlock` (on linux at least), and optimized to avoid
delaying a writer lock for too long, to avoid to much overhead to GC
collections.

Requires BDWGC to be patched with `bdwgc-set-stackbottom.patch` to
be capable to declare the stack bottom of any thread.

ysbaddaden committed:
Use Atomic::Flag to setup Signal handlers

ysbaddaden committed:
Add Thread::RWLock

ysbaddaden committed:
MT: safe channels (but disables select)

Channel now uses Crystal::Mutex and Crystal::ConditionVariable. It's
slow but safe.

`select` is disabled because mutex & condition variables assume that
a fiber can only be in a single waiting queue at a time (otherwise
it's suspended), but `select` expects a single fiber to be queued
into many wait lists, which is impossible.

ysbaddaden committed:
MT: stealing schedulers

Based on "Scheduling Multithreaded Computations by Work Stealing"
(2001) by Nimar S. Arora, Robert D. Blumofe and C. Greg Plaxton, as
well as "Verification of a Concurrent Deque Implementation" (1999)
by Robert D. Blumofe, C. Greg Plaxton and Sandip Ray.

ysbaddaden committed:
MT: manually initialize core structures

Creates a crystal core foundation initializers design.

Just like Crystal's main calls `GC.init` to initialize some static C
variables, this patch introduces some `init` class methods on core
foundation data structures that must be initialized before anything
else (e.g. Thread, Fiber, EventLoop, Scheduler, Signal, Hasher, ...).

ysbaddaden committed:
MT: safe event loop

Enables a nonblocking run of the libevent base loop.

Enqueues IO read/write events, instead of resuming immediately, so a
scheduler can fill its queue and let other schedulers steal from it,
which shall avoid thread contention on running the event loop.

ysbaddaden committed:
Main: execute user code in dedicated fiber

ysbaddaden committed:
MT: main starts threads & schedulers

Starts NPROCS threads, each running their own stealing scheduler.

Runs the main user code (including core/stdlib initializations) in a
dedicated fiber.

Keeps the main thread pending, until the main user code fiber has
terminated; use it to clean the fiber stack pool regularly, and
eventually to receive signals.

ysbaddaden committed:
MT: safe fiber stack pool

ysbaddaden committed:
MT: eventually park threads with empty runnables queue

Avoids starving threads to burn CPU while spinning like crazy to try
and steal runnable fibers when there aren't anything to steal. For
example when waiting on the event loop.

Keeps a counter to never park the last standing thread, but to make
it wait on the event loop instead.

ysbaddaden committed:
MT: rework Crystal::EventLoop and IO::Evented

EventLoop now uses a single Event per Fiber, and declares a little
more generic API: `wait(io, :read | :write, timeout)` and
`sleep(time)`.

IO::Evented now initializes the Fiber Event for each Fiber that is
waiting for a file descriptor to be readable or writable (libevent2
allows multiple events to the same file descriptor), instead of
having a single event, which avoids an issue where concurrent fibers
could alter `#read_timeout` or `#write_timeout` and affect
previously set events.

IO::Evented is now thread-safe, and doesn't use a Deque anymore, but
Crystal::WaitDeque, a double-ended linked list of pending fibers —we
still need to keep a list of pending fibers to resume when an IO is
reopened or closed (EV_CLOSED is only available for the linux/epoll
backend).

The `sleep` methods now delegate to EventLoop, instead of Scheduler.

FIXME: Crystal::Event is kept as a struct to avoid a double
memory allocation (one for Crystal::Event and another one for
LibEvent2::Event) despite having ivars that can be modified from
different places which requires to have a direct access to the
struct through hackish `fiber.@event.as(Crystal::Event)` means. This
should be refactored, for example to have a single Crystal::Event
allocation with enough space to initialize a LibEvent2::Event field.

ysbaddaden committed:
Workaround for bug #6920

The Fiber <-> EventLoop <-> Scheduler circular dependencies
regularly causes bug #6920 to happen. See
https://github.com/crystal-lang/crystal/issues/6920 for details.

Using `dprintf(3)` avoids this inception issue, until the bug is
fixed.

asterite commented:
It would be really cool if you and @bcardiff combined efforts on this topic.

From my point of view, I think it would be cool to have any safe infrastructure for running `spawn` across multiple threads, even if it's slower than now (for example using `GC.enable` and `GC.disable`). Once that's done std code can be upgraded to be multithread-safe, and also work can be done to improve the performance. In the meantime also libraries can be upgraded to consider multithreading and so on. So in a way this would be parallelizing this feature's implementation :-)

bararchy commented:
Its quite funny , I was 100% that you are both synced on the releases hahah

bcardiff commented:
Can we get https://github.com/crystal-lang/crystal/pull/7547/commits/2e19060fb007336a9002048297b6d3af9829d8be and https://github.com/crystal-lang/crystal/pull/7547/commits/491bdaea37d91d44049d66b3bc055fd38860fc87 in their own PRs?

I think we can merge those right away.

Note that https://github.com/crystal-lang/crystal/pull/7547/commits/35d30af3e05dc56e6d1d4611b2ebece0ec7368ad is already in master.

Sija commented on spec/std/channel_spec.cr:
```diff
@@ -5,6 +5,14 @@ private def yield_to(fiber)
+  {% if flag?(:mt) %}
+    pending({{name}}) { {{yield}} }
+  {% else %}
+    describe({{name}}) { {{yield}} }
```
Why not `it`?

ysbaddaden replied:
Typo :)

ysbaddaden commented:
@asterite this is why I pushed the PR instead of keeping it hidden in some branch :)

I'm not aiming to merge for now.

I'm disappointed by the performance of MT for the time being. I'm starting to have acceptable single thread performance, despite being thread safe (overhead is contained) which is a nice achievement, but I don't get any improvement with more threads... channels are awfully slower, and select is disabled —I also broke IO::Evented but that's fixable.

The best performance is currently achieved with current master in "sharding" (distinct processes) and no GC, which has a higher impact on performance than I thought (actual collections, not the safe context switches).

I'll try a Chase and Lev queue for the scheduler. It would avoid a mmap and use a circular queue (that could grow and shrink), and allow to steal many fibers are once, instead of stealing only one at a time.

ysbaddaden commented:
@bcardiff extracted and rebased.

stakach commented:
Not entirely sure if this is the place to discuss this, however I feel like this might be the wrong direction to take the language. Fibers are awesome, they are quite mutation safe, very easy to reason with.

Might it not be better for the language to embrace single thread IO reactor with a thread pool accessible via something like a `ThreadChannel` for processor intensive operations that don't have access to IO operations.

I personally really like the look of #6468 and it feels like the complexity of managing fibers will go up once threads are introduced - given fibers should always run on the same thread.

In short, processes feel like the right way to achieve parallelism however it would be nice to perform some processor intensive operations without blocking the reactor.

ysbaddaden commented:
#6461 is unrelated, there can be structured concurrency with MT and stealing schedulers.

The complexity you refer to is fearless concurrency which is already a problem, even without threads, it's just not as obvious. This should have it's own discussion. I'll try to open it during the weekend with some ideas (initialize constants at start time, thread safe globals only, *moving* values accross fibers) and links to interesting articles.

As stated above, I'm disappointed with the performance, but I didn't try to add an actor to verify a bcrypt hash, or to calculate a random number in fibers, those cases should be dramatically faster the more threads we add (limited by hardware).

ysbaddaden committed:
Benchmarks: switch, channel, mutex and http server

The switch, channel and mutex benchmarks are furious benchmarks that
will report and allow to compare the overhead of synchronization
primitives. They must be taken with a grain of salt, as they don't
represent any real world constraint.

Note that Fiber.yield with the current scheduler goes through the
event loop, which is not the case for the MT scheduler which
immediately switches fibers, so the switch benchmark is *much*
faster with MT and the actual difference with ST is the libevent
overhead!

The http server test is more interesting since it spawns fibers,
allocates memory (thus runs GC collections), involves the event loop
(thus libevent), and does some IO. It's still not representative of
a real world application, but gives a nice overview of the overhead
of MT, and the actual impact of changing or tweaking some internals.

NOTE: since Socket#accept is broken because of broken IO::Evented,
the http server benchmark doesn't currently run with MT and the
server will just hang.

ysbaddaden committed:
fixup: add Fiber::Context#can_resume?

ysbaddaden commented:
Sigh. Benchmarking is tricky.

With default `wrk` settings (2 threads, 10 connections), running 2 or more threads isn't better than 1 (unless requests are heavily CPU bound) and slower than current master, which is *very* disappointing. But `wrk` with 1 thread and 128 shows something completely different. Using relative results:

```
single-thread 1 process:   × 1.00
single-thread 2 processes: × 1.35
single-thread 3 processes: × 1.65
```
```
multi-thread 1 thread:     × 1.10
multi-thread 2 threads:    × 1.50
multi-thread 3 threads:    × 1.40
```

With enough incoming connections, MT is starting to get faster then running different processes :fireworks:, yet it's still slower with 3+ threads (too much contention? scheduler? gc?).

ysbaddaden committed:
CI: specs are broken, don't clog circle/travis with useless runs [ci skip]

ysbaddaden committed:
Add Thread.log with per-thread color for MT debug

ysbaddaden committed:
fixup: GC: try out different thread safety strategies

wontruefree commented on src/concurrent/mt_scheduler.cr:
```diff
@@ -0,0 +1,304 @@
+  protected def steal_loop : Fiber?
+    Thread.log "scheduler#steal_loop", self
+
+    100.times do
```
I am just curious why 100?
This seems like a high number but I do not do much MT programming.

ysbaddaden replied:
Magic number. It should probably be much higher, to avoid parking the thread to quickly, while enqueue should wait for the queue to have some entries before unparking a thread, to make it worth it.

ysbaddaden replied:
Maybe it should be adaptive, too, depending on how much time was spent waiting in the previous steal loop, and whether the thread was parked or not.

cpunion commented:
Great! How to build and test this?

asterite commented:
I think this can be closed now that #8112 was merged.

@ysbaddaden Sorry this didn't make it through!
